Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java

```diff
@@ -0,0 +1,142 @@
+import lombok.SneakyThrows;
+
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+public class AzurePerformanceMetricsCollector {
+
+    public static final String NONE = "NONE";
+    public static final String PERFORMANCE = "PERFORMANCE";
+
+    public static void beforeEnter(final AzureOperationRef operation) {
+        final Deque<AzureOperationRef> operations = AzureTaskContext.getContextOperations();
+        AzurePerformanceMetricsCollector.enterOperation(operation, operations);
+    }
+
+    public static void afterExit(final AzureOperationRef operation) {
+        final Deque<AzureOperationRef> operations = AzureTaskContext.getContextOperations();
+        AzurePerformanceMetricsCollector.exitOperation(operation, operations);
+    }
+
+    public static void beforeEnter(final AzureTask<?> task) {
+        AzurePerformanceMetricsCollector.enterTask(task);
+    }
+
+    public static void afterExit(final AzureTask<?> task) {
+        AzurePerformanceMetricsCollector.exitTask(task);
+        final AzureTaskContext.Node context = task.getContext();
+        final Deque<AzureOperationRef> operations = AzureTaskContext.getContextOperations(context);
+        while (!operations.isEmpty()) { // end operations asynchronously
+            final AzureOperationRef current = operations.pop();
+            AzurePerformanceMetricsCollector.endOperation(current, operations);
+            final AzureOperation annotation = AzureOperationUtils.getAnnotation(current);
+            if (annotation.type() == AzureOperation.Type.ACTION) {
+                break; // should not end side-effect action.
+            }
+        }
+    }
+
+    private static void enterOperation(final AzureOperationRef current, final Deque<? extends AzureOperationRef> ancestors) {
+        final long timestamp = System.currentTimeMillis();
+        final HashMap<String, String> properties = buildOperationProperties(current, ancestors);
+        properties.put("timestamp", String.valueOf(timestamp));
+        properties.put("action", "ENTER");
+        sendTelemetry(properties);
+    }
+
+    private static void exitOperation(final AzureOperationRef current, final Deque<? extends AzureOperationRef> ancestors) {
+        final long timestamp = System.currentTimeMillis();
+        final HashMap<String, String> properties = buildOperationProperties(current, ancestors);
+        properties.put("timestamp", String.valueOf(timestamp));
+        properties.put("action", "EXIT");
+        sendTelemetry(properties);
+    }
+
+    private static void endOperation(final AzureOperationRef current, final Deque<? extends AzureOperationRef> ancestors) {
+        final long timestamp = System.currentTimeMillis();
+        final HashMap<String, String> properties = buildOperationProperties(current, ancestors);
+        properties.put("timestamp", String.valueOf(timestamp));
+        properties.put("action", "END");
```

Comment:
action values: enter end? give a better name to replace "action"

Reply:
remove, we now only have `ENTER`/`EXIT`